server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java

```diff
@@ -91,6 +93,7 @@ public class SinkQuerySegmentWalker implements QuerySegmentWalker
   private final String dataSource;
 
   private final VersionedIntervalTimeline<String, Sink> sinkTimeline;
+  private final VersionedIntervalTimeline<String, SegmentDescriptorOvershadowable> upgradedSegmentsTimeline;
```

Comment:
Do we need all three of `sinkTimeline`, `upgradeDescriptorTimeline`, and `newIdToBasePendingSegment`? Seems like we should be able to do with just the upgraded timeline, where the objects in the upgraded timeline link back to the proper Sink. Perhaps that'd be easier to keep in sync.

Reply:
I had thought about it but was unsure about adding an entry like
<newVersion, newInterval, sink> to a VersionedIntervalTimeline as it seems to have some logic tying newVersion and sink (with a different version).

I had also considered creating a new sink but that might be incorrect as ingestion may not work correctly.

Reply:
I've taken the approach of having a single a timeline as suggested, and added more tests.